File: Swift/sorting/mergeSort.swift
/*
 * 1. Brief description of the algorithm
 * Merge Sort is a highly efficient, stable, comparison-based sorting algorithm.
 * It's a classic example of the "divide and conquer" paradigm that works by
 * recursively splitting the array into halves, sorting them, and then merging them back.
 *
 * 2. Time complexity: O(n log n)
 * This is consistent for the best, average, and worst cases because the array is
 * always divided in half.
 *
 * 3. Space complexity: O(n)
 * This is because it requires additional memory to store the merged subarrays.
 */

// 4. Implementation
func mergeSort(_ array: [Int]) -> [Int] {
    // Base case: An array with zero or one element is already sorted.
    guard array.count > 1 else {
        return array
    }

    // Divide the array into two halves.
    let middleIndex = array.count / 2
    
    // Recursively sort the left and right halves.
    let leftArray = mergeSort(Array(array[0..<middleIndex]))
    let rightArray = mergeSort(Array(array[middleIndex..<array.count]))
    
    // Conquer by merging the two sorted halves.
    return merge(leftPile: leftArray, rightPile: rightArray)
}

func merge(leftPile: [Int], rightPile: [Int]) -> [Int] {
    var leftIndex = 0
    var rightIndex = 0
    var orderedPile: [Int] = []
    orderedPile.reserveCapacity(leftPile.count + rightPile.count)

    // Compare elements from both piles and append the smaller one.
    while leftIndex < leftPile.count && rightIndex < rightPile.count {
        if leftPile[leftIndex] < rightPile[rightIndex] {
            orderedPile.append(leftPile[leftIndex])
            leftIndex += 1
        } else {
            orderedPile.append(rightPile[rightIndex])
            rightIndex += 1
        }
    }

    // Append any remaining elements.
    orderedPile.append(contentsOf: leftPile[leftIndex...])
    orderedPile.append(contentsOf: rightPile[rightIndex...])

    return orderedPile
}

// 5. Example usage/test cases
let unsortedArray = [38, 27, 43, 3, 9, 82, 10]
print("Unsorted Array: \(unsortedArray)")
let sortedArray = mergeSort(unsortedArray)
print("Sorted Array  : \(sortedArray)") // Expected: [3, 9, 10, 27, 38, 43, 82]

let anotherArray = [5, 1, 4, 8, 2, 9, 3, 7, 6]
print("\nUnsorted Array: \(anotherArray)")
let sortedAnotherArray = mergeSort(anotherArray)
print("Sorted Array  : \(sortedAnotherArray)") // Expected: [1, 2, 3, 4, 5, 6, 7, 8, 9]

File: C#/sorting/MergeSort.cs
using System;

/// <summary>
/// Merge Sort using divide and conquer with auxiliary array.
/// </summary>
public static class MergeSort
{
    // Time: O(n log n), Space: O(n)
    public static void Sort(int[] arr)
    {
        if (arr == null || arr.Length <= 1) return;
        int[] aux = new int[arr.Length];
        SortRange(arr, aux, 0, arr.Length - 1);
    }

    private static void SortRange(int[] arr, int[] aux, int left, int right)
    {
        if (left >= right) return;
        int mid = left + ((right - left) / 2);
        SortRange(arr, aux, left, mid);
        SortRange(arr, aux, mid + 1, right);
        Merge(arr, aux, left, mid, right);
    }

    private static void Merge(int[] arr, int[] aux, int left, int mid, int right)
    {
        int i = left, j = mid + 1, k = left;
        while (i <= mid && j <= right)
        {
            if (arr[i] <= arr[j]) aux[k++] = arr[i++]; else aux[k++] = arr[j++];
        }
        while (i <= mid) aux[k++] = arr[i++];
        while (j <= right) aux[k++] = arr[j++];
        for (int t = left; t <= right; t++) arr[t] = aux[t];
    }

    public static void Main()
    {
        int[] arr = { 38, 27, 43, 3, 9, 82, 10 };
        Sort(arr);
        Console.WriteLine("[MergeSort] Sorted: " + string.Join(", ", arr));
    }
}


